File: DotNetCode/Controllers/LeaveController.cs
using OfficeConnect_Web.Models;
using OfficeConnect_Web.ViewModel;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Web;
using System.Web.Mvc;
using static OfficeConnect_Web.Models.LeaveModel;

namespace OfficeConnect_Web.Controllers
{
    [AuthAttribute]
    public class LeaveController : Controller
    {
        LeaveModel LM = new LeaveModel();

        // POST: Leave/GetAllLeaveType
        [Route("Leave/GetAllLeaveType")]
        [HttpPost]
        public ActionResult GetAllLeaveType(LeaveTypeViewModel model)
        {
            try
            {
                var leave = LM.GetAllLeaveType(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/GetLeaveType
        [Route("Leave/GetLeaveType")]
        [HttpPost]
        public ActionResult GetLeaveType(LeaveTypeViewModel model)
        {
            try
            {
                var leave = LM.GetLeaveType(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/AddLeaveType
        [Route("Leave/AddLeaveType")]
        [HttpPost]
        public ActionResult AddLeaveType(LeaveTypeViewModel model)
        {
            try
            {
                var leave = LM.AddLeaveType(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/UpdateLeaveType
        [Route("Leave/UpdateLeaveType")]
        [HttpPost]
        public ActionResult UpdateLeaveType(LeaveTypeViewModel model)
        {
            try
            {
                var leave = LM.UpdateLeaveType(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/DeleteLeaveType
        [Route("Leave/DeleteLeaveType")]
        [HttpPost]
        public ActionResult DeleteLeaveType(LeaveTypeViewModel model)
        {
            try
            {
                var leave = LM.DeleteLeaveType(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/ActivateLeaveType
        [Route("Leave/ActivateLeaveType")]
        [HttpPost]
        public ActionResult ActivateLeaveType(LeaveTypeViewModel model)
        {
            try
            {
                var leave = LM.ActivateLeaveType(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/DeactivateLeaveType
        [Route("Leave/DeactivateLeaveType")]
        [HttpPost]
        public ActionResult DeactivateLeaveType(LeaveTypeViewModel model)
        {
            try
            {
                var leave = LM.DeactivateLeaveType(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/DDLeaveType
        [Route("Leave/DDLeaveType")]
        [HttpPost]
        public ActionResult DDLeaveType(DDLeaveTypePayloadViewModel model)
        {
            try
            {
                var leave = LM.GetDDLeaveType(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/DDApproveManager
        [Route("Leave/DDApproveManager")]
        [HttpPost]
        public ActionResult DDApproveManager(DDComOffManager model)
        {
            try
            {
                var leave = LM.DDApproveManager(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/CompOffLeave
        [Route("Leave/CompOffLeave")]
        [HttpPost]
        public ActionResult CompOffLeave(CompOffRequestViewModel model)
        {
            try
            {
                var leave = LM.CompOffLeave(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/CompOffHours
        [Route("Leave/CompOffHours")]
        [HttpPost]
        public ActionResult CompOffHours(CompOffRequestViewModel model)
        {
            try
            {
                var leave = LM.CompOffHours(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/GetAllEmpCompOffLeave
        [Route("Leave/GetAllEmpCompOffLeave")]
        [HttpPost]
        public ActionResult GetAllEmpCompOffLeave(CompOffRequestViewModel model)
        {
            try
            {
                var leave = LM.GetAllEmpCompOffLeave(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/GetAllCompOffLeave
        [Route("Leave/GetAllCompOffLeave")]
        [HttpPost]
        public ActionResult GetAllCompOffLeave(CompOffRequestViewModel model)
        {
            try
            {
                var leave = LM.GetAllCompOffLeave(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/ApproveCompOff
        [Route("Leave/ApproveCompOff")]
        [HttpPost]
        public ActionResult ApproveCompOff(ApproveCompOffViewModel model)
        {
            try
            {
                var leave = LM.ApproveCompOff(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/RejectCompOff
        [Route("Leave/RejectCompOff")]
        [HttpPost]
        public ActionResult RejectCompOff(ApproveCompOffViewModel model)
        {
            try
            {
                var leave = LM.RejectCompOff(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/DraftLeave
        [Route("Leave/DraftLeave")]
        [HttpPost]
        public ActionResult DraftLeave(EmpLeaveApplicationViewModel model)
        {
            try
            {
                var leave = LM.DraftLeave(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/ApplyLeave
        [Route("Leave/ApplyLeave")]
        [HttpPost]
        public ActionResult ApplyLeave(EmpLeaveApplicationViewModel model)
        {
            try
            {
                var leave = LM.ApplyLeave(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/DraftApplyLeave
        [Route("Leave/DraftApplyLeave")]
        [HttpPost]
        public ActionResult DraftApplyLeave(EmpLeaveApplicationViewModel model)
        {
            try
            {
                var leave = LM.DraftApplyLeave(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/WithDrawLeave
        [Route("Leave/WithDrawLeave")]
        [HttpPost]
        public ActionResult WithDrawLeave(EmpLeaveApplicationViewModel model)
        {
            try
            {
                var leave = LM.WithDrawLeave(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/CancelLeave
        [Route("Leave/CancelLeave")]
        [HttpPost]
        public ActionResult CancelLeave(EmpLeaveApplicationViewModel model)
        {
            try
            {
                var leave = LM.CancelLeave(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/DeleteDraftLeave
        [Route("Leave/DeleteDraftLeave")]
        [HttpPost]
        public ActionResult DeleteDraftLeave(EmpLeaveApplicationViewModel model)
        {
            try
            {
                var leave = LM.DeleteDraftLeave(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/GetAllLeave
        [Route("Leave/GetAllLeave")]
        [HttpPost]
        public ActionResult GetAllLeave(EmpLeaveApplicationViewModel model)
        {
            try
            {
                var leave = LM.GetAllLeave(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/GetAllApplyManagerLeave
        [Route("Leave/GetAllApplyManagerLeave")]
        [HttpPost]
        public ActionResult GetAllApplyManagerLeave(EmpLeaveApplicationViewModel model)
        {
            try
            {
                var leave = LM.GetAllApplyManagerLeave(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/GetAllManagerLeave
        [Route("Leave/GetAllManagerLeave")]
        [HttpPost]
        public ActionResult GetAllManagerLeave(EmpLeaveApplicationViewModel model)
        {
            try
            {
                var leave = LM.GetAllManagerLeave(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/ApproveLeaveByManager
        [Route("Leave/ApproveLeaveByManager")]
        [HttpPost]
        public ActionResult ApproveLeaveByManager(ApproveLeaveViewModel model)
        {
            try
            {
                var leave = LM.ApproveLeaveByManager(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/RejectLeaveByManager
        [Route("Leave/RejectLeaveByManager")]
        [HttpPost]
        public ActionResult RejectLeaveByManager(ApproveLeaveViewModel model)
        {
            try
            {
                var leave = LM.RejectLeaveByManager(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/GetAllApplyHRLeave
        [Route("Leave/GetAllApplyHRLeave")]
        [HttpPost]
        public ActionResult GetAllApplyHRLeave(EmpLeaveApplicationViewModel model)
        {
            try
            {
                var leave = LM.GetAllApplyHRLeave(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/GetAllHRLeave
        [Route("Leave/GetAllHRLeave")]
        [HttpPost]
        public ActionResult GetAllHRLeave(EmpLeaveApplicationViewModel model)
        {
            try
            {
                var leave = LM.GetAllHRLeave(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/ApproveLeaveByHR
        [Route("Leave/ApproveLeaveByHR")]
        [HttpPost]
        public ActionResult ApproveLeaveByHR(ApproveLeaveViewModel model)
        {
            try
            {
                var leave = LM.ApproveLeaveByHR(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/RejectLeaveByHR
        [Route("Leave/RejectLeaveByHR")]
        [HttpPost]
        public ActionResult RejectLeaveByHR(ApproveLeaveViewModel model)
        {
            try
            {
                var leave = LM.RejectLeaveByHR(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/UploadFileLeave
        [Route("Leave/UploadFileLeave")]
        [HttpPost]
        public ActionResult UploadFileLeave(FileUploadAPIViewModel model)
        {
            try
            {
                var leave = LM.UploadFileLeave(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/IndividualLeaveCount
        [Route("Leave/IndividualLeaveCount")]
        [HttpPost]
        public ActionResult IndividualLeaveCount(EmpLeaveApplicationViewModel model)
        {
            try
            {
                var leave = LM.IndividualLeaveCount(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
        // POST: Leave/LeaveBalReport
        [Route("Leave/LeaveBalReport")]
        [HttpPost]
        public ActionResult LeaveBalReport(LeaveBalReportViewModel model)
        {
            try
            {
                var leave = LM.LeaveBalReport(model);
                return Json(leave, JsonRequestBehavior.AllowGet);
            }
            catch (CustomApiException ex)
            {
                return Json(new { StatusCode = ex.StatusCode, Message = ex.Message });
            }
        }
    }
}

File: java/src/main/java/com/officeconnect/controller/LeaveController.java
package com.officeconnect.controller;

import com.officeconnect.dto.*;
import com.officeconnect.service.LeaveService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/Leave")
public class LeaveController {

    @Autowired
    private LeaveService leaveService;

    @PostMapping("/GetAllLeaveType")
    public ResponseEntity<?> getAllLeaveType(@RequestBody LeaveTypeViewModel model) {
        try {
            List<LeaveTypeViewModel> result = leaveService.getAllLeaveType(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/GetAllLeaveTypes")
    public ResponseEntity<?> getAllLeaveTypes(@RequestBody LeaveTypeViewModel model) {
        try {
            List<LeaveTypeViewModel> result = leaveService.getAllLeaveType(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/DDLeaveType")
    public ResponseEntity<?> ddLeaveType(@RequestBody DDLeaveTypePayloadViewModel model) {
        try {
            List<LeaveTypeViewModel> result = leaveService.getDDLeaveType(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/AddLeaveType")
    public ResponseEntity<?> addLeaveType(@RequestBody LeaveTypeViewModel model) {
        try {
            LeaveTypeViewModel result = leaveService.addLeaveType(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/UpdateLeaveType")
    public ResponseEntity<?> updateLeaveType(@RequestBody LeaveTypeViewModel model) {
        try {
            LeaveTypeViewModel result = leaveService.updateLeaveType(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/DeleteLeaveType")
    public ResponseEntity<?> deleteLeaveType(@RequestBody LeaveTypeViewModel model) {
        try {
            LeaveTypeViewModel result = leaveService.deleteLeaveType(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/ApplyLeave")
    public ResponseEntity<?> applyLeave(@RequestBody EmpLeaveApplicationViewModel model) {
        try {
            EmpLeaveApplicationViewModel result = leaveService.applyLeave(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.ok(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/DraftLeave")
    public ResponseEntity<?> draftLeave(@RequestBody EmpLeaveApplicationViewModel model) {
        try {
            EmpLeaveApplicationViewModel result = leaveService.draftLeave(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/GetAllLeave")
    public ResponseEntity<?> getAllLeave(@RequestBody EmpLeaveApplicationViewModel model) {
        try {
            List<EmpLeaveApplicationViewModel> result = leaveService.getAllEmpLeave(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/GetAllManagerLeave")
    public ResponseEntity<?> getAllManagerLeave(@RequestBody EmpLeaveApplicationViewModel model) {
        try {
            List<EmpLeaveApplicationViewModel> result = leaveService.getAllManagerLeave(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/ApproveLeaveByManager")
    public ResponseEntity<?> approveLeaveByManager(@RequestBody ApproveLeaveViewModel model) {
        try {
            ApproveLeaveViewModel result = leaveService.approveLeaveByManager(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/ApproveLeaveByHR")
    public ResponseEntity<?> approveLeaveByHR(@RequestBody ApproveLeaveViewModel model) {
        try {
            ApproveLeaveViewModel result = leaveService.approveLeaveByHR(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/CompOffLeave")
    public ResponseEntity<?> compOffLeave(@RequestBody CompOffRequestViewModel model) {
        try {
            CompOffRequestViewModel result = leaveService.compOffLeave(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/ApproveCompOff")
    public ResponseEntity<?> approveCompOff(@RequestBody ApproveLeaveViewModel model) {
        try {
            ApproveLeaveViewModel result = leaveService.approveCompOff(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/WithDrawLeave")
    public ResponseEntity<?> withDrawLeave(@RequestBody EmpLeaveApplicationViewModel model) {
        try {
            EmpLeaveApplicationViewModel result = leaveService.draftLeave(model);
            result.setStatus("Withdrawn");
            result.setMsg("Leave withdrawn successfully");
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/CancelLeave")
    public ResponseEntity<?> cancelLeave(@RequestBody EmpLeaveApplicationViewModel model) {
        try {
            Map<String, Object> result = leaveService.cancelLeave(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/ActivateLeaveType")
    public ResponseEntity<?> activateLeaveType(@RequestBody LeaveTypeViewModel model) {
        try {
            LeaveTypeViewModel result = leaveService.activateLeaveType(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/DeactivateLeaveType")
    public ResponseEntity<?> deactivateLeaveType(@RequestBody LeaveTypeViewModel model) {
        try {
            LeaveTypeViewModel result = leaveService.deactivateLeaveType(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/RejectCompOff")
    public ResponseEntity<?> rejectCompOff(@RequestBody ApproveLeaveViewModel model) {
        try {
            ApproveLeaveViewModel result = leaveService.rejectCompOff(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/DraftApplyLeave")
    public ResponseEntity<?> draftApplyLeave(@RequestBody EmpLeaveApplicationViewModel model) {
        try {
            EmpLeaveApplicationViewModel result = leaveService.draftApplyLeave(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/DDApproveManager")
    public ResponseEntity<?> ddApproveManager(@RequestBody Map<String, Object> model) {
        try {
            List<Map<String, Object>> result = leaveService.getDDApproveManager();
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/CompOffHours")
    public ResponseEntity<?> compOffHours(@RequestBody CompOffRequestViewModel model) {
        try {
            CompOffRequestViewModel result = leaveService.getCompOffHours(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/GetAllEmpCompOffLeave")
    public ResponseEntity<?> getAllEmpCompOffLeave(@RequestBody CompOffRequestViewModel model) {
        try {
            List<CompOffRequestViewModel> result = leaveService.getAllEmpCompOffLeave(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/GetAllCompOffLeave")
    public ResponseEntity<?> getAllCompOffLeave(@RequestBody CompOffRequestViewModel model) {
        try {
            List<CompOffRequestViewModel> result = leaveService.getAllCompOffLeave(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/GetLeaveType")
    public ResponseEntity<?> getLeaveType(@RequestBody LeaveTypeViewModel model) {
        try {
            LeaveTypeViewModel result = leaveService.getLeaveType(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/DeleteDraftLeave")
    public ResponseEntity<?> deleteDraftLeave(@RequestBody EmpLeaveApplicationViewModel model) {
        try {
            EmpLeaveApplicationViewModel result = leaveService.deleteDraftLeave(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/GetAllApplyManagerLeave")
    public ResponseEntity<?> getAllApplyManagerLeave(@RequestBody EmpLeaveApplicationViewModel model) {
        try {
            List<EmpLeaveApplicationViewModel> result = leaveService.getAllApplyManagerLeave(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/RejectLeaveByManager")
    public ResponseEntity<?> rejectLeaveByManager(@RequestBody ApproveLeaveViewModel model) {
        try {
            ApproveLeaveViewModel result = leaveService.rejectLeaveByManager(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/GetAllApplyHRLeave")
    public ResponseEntity<?> getAllApplyHRLeave(@RequestBody EmpLeaveApplicationViewModel model) {
        try {
            List<EmpLeaveApplicationViewModel> result = leaveService.getAllApplyHRLeave(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/GetAllHRLeave")
    public ResponseEntity<?> getAllHRLeave(@RequestBody EmpLeaveApplicationViewModel model) {
        try {
            List<EmpLeaveApplicationViewModel> result = leaveService.getAllHRLeave(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/IndividualLeaveCount")
    public ResponseEntity<?> individualLeaveCount(@RequestBody EmpLeaveApplicationViewModel model) {
        try {
            Map<String, Object> result = leaveService.getIndividualLeaveCount(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/LeaveBalReport")
    public ResponseEntity<?> leaveBalReport(@RequestBody Map<String, Object> model) {
        try {
            List<Map<String, Object>> result = leaveService.leaveBalReport(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/RejectLeaveByHR")
    public ResponseEntity<?> rejectLeaveByHR(@RequestBody ApproveLeaveViewModel model) {
        try {
            ApproveLeaveViewModel result = leaveService.rejectLeaveByHR(model);
            return ResponseEntity.ok(result);
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("StatusCode", 404, "Message", ex.getMessage()));
        }
    }

    @PostMapping("/UploadFileLeave")
    public ResponseEntity<?> uploadFileLeave(@RequestParam("file") MultipartFile file, 
                                              @RequestParam("empId") String empId,
                                              @RequestParam("leaveTypeId") String leaveTypeId) {
        try {
            // Process the uploaded file
            String fileName = file.getOriginalFilename();
            byte[] fileContent = file.getBytes();
            
            // TODO: Implement file processing logic here
            // For now, just return success
            return ResponseEntity.ok(Map.of("StatusCode", 200, "Message", "File uploaded successfully: " + fileName));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(Map.of("StatusCode", 500, "Message", "File upload failed: " + e.getMessage()));
        }
    }
}